virtcontainers/pkg/cgroups/manager.go

```diff
@@ -0,0 +1,313 @@
+// Copyright (c) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package cgroups
+
+import (
+	"bufio"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/kata-containers/runtime/virtcontainers/pkg/rootless"
+	libcontcgroups "github.com/opencontainers/runc/libcontainer/cgroups"
+	libcontcgroupsfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
+	libcontcgroupssystemd "github.com/opencontainers/runc/libcontainer/cgroups/systemd"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"github.com/opencontainers/runc/libcontainer/specconv"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+type Config struct {
+	// Cgroups specifies specific cgroup settings for the various subsystems that the container is
+	// placed into to limit the resources the container has available
+	// If nil, New() will create one.
+	Cgroups *configs.Cgroup
+
+	// CgroupPaths contains paths to all the cgroups setup for a container. Key is cgroup subsystem name
+	// with the value as the path.
+	CgroupPaths map[string]string
+
+	// Resources represents the runtime resource constraints
+	Resources specs.LinuxResources
+
+	// CgroupPath is the OCI spec cgroup path
+	CgroupPath string
+}
+
+type Manager struct {
+	sync.Mutex
+	mgr libcontcgroups.Manager
+}
+
+const (
+	// file in the cgroup that contains the pids
+	cgroupProcs = "cgroup.procs"
+)
+
+var (
+	// If set to true, expects cgroupsPath to be of form "slice:prefix:name", otherwise cgroups creation will fail
+	systemdCgroup *bool
+
+	cgroupsLogger = logrus.WithField("source", "virtcontainers/pkg/cgroups")
+)
+
+func EnableSystemdCgroup() {
+	systemd := true
+	systemdCgroup = &systemd
+}
+
+func UseSystemdCgroup() bool {
+	if systemdCgroup != nil {
+		return *systemdCgroup
+	}
+	return false
+}
+
+// returns the list of devices that a hypervisor may need
+func hypervisorDevices() []specs.LinuxDeviceCgroup {
+	wildcard := int64(-1)
+
+	devices := []specs.LinuxDeviceCgroup{}
+
+	devices = append(devices,
+		// hypervisor needs access to all devicemapper devices, otherwise
+		// the device cgroups MUST BE updated when a new container is added to the POD.
+		specs.LinuxDeviceCgroup{
+			Allow:  true,
+			Type:   "b",
+			Major:  &wildcard,
+			Minor:  &wildcard,
+			Access: "rwm",
+		})
+
+	// hypervisor may need access to these devices
+	hypervisorDevices := []string{"/dev/kvm", "/dev/vhost-net"}
```

Comment:
I assume this is needed just for cgroupv2? Its better to clarify it in the comments here.